### ISSUE TITLE
Blockchain will not download

### ISSUE THREAD

**Issue by PavelNiedoba** [bitcoinj/bitcoinj#1240]
I was using bitcoinj wallet sucessfully since February. It was "quite stable". Every 2-4 days Blockchain class stopped downloading new blocks.  Simple restart solved the problem. Today morning the wallet was stuck again. Confirmed balance was 0, unconfirmed some positive number.  Restart didn't work this time.  I deleted wallet and spv blockchain files and restored wallet from seed. The blockchain downloaded 98% of headers then slowed down and stopped completely. Another restart gives it some more blocks and then it freezes again. 

I tried to soft and hard restart multiple times. Always same result. I increased java memory for case this is the problem. I have more than one thousand of transactions in this wallet. I use only 5 addresses, don't use any change addresses.  Internet is very fast (100mbit symetric optical)

I'm using Smapshot 14 from February. Nothing has changed in this code since. (Still same jar) 

Log file is full of exceptions, obviously whole peer group is down:

[PeerGroup Thread] INFO org.bitcoinj.core.PeerSocketHandler - [129.240.12.4]:8333 - java.nio.channels.CancelledKeyException
[PeerGroup Thread] WARN org.bitcoinj.net.ConnectionHandler - Error writing message to connection, closing connection
java.nio.channels.CancelledKeyException
	at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
	at sun.nio.ch.SelectionKeyImpl.interestOps(SelectionKeyImpl.java:77)
	at org.bitcoinj.net.ConnectionHandler.setWriteOps(ConnectionHandler.java:106)
	at org.bitcoinj.net.ConnectionHandler.writeBytes(ConnectionHandler.java:152)
	at org.bitcoinj.core.PeerSocketHandler.sendMessage(PeerSocketHandler.java:90)
	at org.bitcoinj.core.Peer.ping(Peer.java:1533)
	at org.bitcoinj.core.Peer.ping(Peer.java:1524)
	at org.bitcoinj.core.PeerGroup$12.run(PeerGroup.java:1679)
	at com.google.common.util.concurrent.MoreExecutors$ScheduledListeningDecorator$NeverSuccessfulListenableFutureTask.run(MoreExecutors.java:648)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at org.bitcoinj.utils.ContextPropagatingThreadFactory$1.run(ContextPropagatingThreadFactory.java:35)
	at java.lang.Thread.run(Thread.java:745)
[PeerGroup Thread] INFO org.bitcoinj.core.PeerSocketHandler - [129.240.12.4]:8333 - java.nio.channels.CancelledKeyException
[PeerGroup Thread] WARN org.bitcoinj.net.ConnectionHandler - Error writing message to connection, closing connection
java.nio.channels.CancelledKeyException
	at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
	at sun.nio.ch.SelectionKeyImpl.interestOps(SelectionKeyImpl.java:77)
	at org.bitcoinj.net.ConnectionHandler.setWriteOps(ConnectionHandler.java:106)
	at org.bitcoinj.net.ConnectionHandler.writeBytes(ConnectionHandler.java:152)
	at org.bitcoinj.core.PeerSocketHandler.sendMessage(PeerSocketHandler.java:90)
	at org.bitcoinj.core.Peer.ping(Peer.java:1533)
	at org.bitcoinj.core.Peer.ping(Peer.java:1524)
	at org.bitcoinj.core.PeerGroup$12.run(PeerGroup.java:1679)
	at com.google.common.util.concurrent.MoreExecutors$ScheduledListeningDecorator$NeverSuccessfulListenableFutureTask.run(MoreExecutors.java:648)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at org.bitcoinj.utils.ContextPropagatingThreadFactory$1.run(ContextPropagatingThreadFactory.java:35)
	at java.lang.Thread.run(Thread.java:745)
[PeerGroup Thread] INFO org.bitcoinj.core.PeerSocketHandler - [129.240.12.4]:8333 - java.nio.channels.CancelledKeyException
[PeerGroup Thread] WARN org.bitcoinj.net.ConnectionHandler - Error writing message to connection, closing connection
java.nio.channels.CancelledKeyException
	at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
	at sun.nio.ch.SelectionKeyImpl.interestOps(SelectionKeyImpl.java:77)
	at org.bitcoinj.net.ConnectionHandler.setWriteOps(ConnectionHandler.java:106)
	at org.bitcoinj.net.ConnectionHandler.writeBytes(ConnectionHandler.java:152)
	at org.bitcoinj.core.PeerSocketHandler.sendMessage(PeerSocketHandler.java:90)
	at org.bitcoinj.core.Peer.ping(Peer.java:1533)
	at org.bitcoinj.core.Peer.ping(Peer.java:1524)
	at org.bitcoinj.core.PeerGroup$12.run(PeerGroup.java:1679)
	at com.google.common.util.concurrent.MoreExecutors$ScheduledListeningDecorator$NeverSuccessfulListenableFutureTask.run(MoreExecutors.java:648)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)

**Comment by schildbach** [comment]
These exception look like connection problems.

**Comment by PavelNiedoba** [comment]
Well, there is connection problem for sure, but not caused by internet connection.  It's happening on hosted server, there are web services running 100% reliably latency is under 1 ms. 
More than that it will happen also in my office with 20mbit adsl.

**Comment by schildbach** [comment]
Can you try on current master? If the problem persists, can you upload the full log?

**Comment by PavelNiedoba** [comment]
Yes sir. 

  I'm on master head now. The problem persist. It looks like every time transaction is found in incomming block, current peer gets closed:
2016-04-12 15:05:55 [NioClientManager] INFO  o.b.c.l.DownloadProgressTracker - Chain download switched to [78.106.74.9]:8333
2016-04-12 15:05:55 [NioClientManager] WARN  org.bitcoinj.net.ConnectionHandler - Error handling SelectionKey: java.lang.IllegalStateException
2016-04-12 15:05:56 [Wallet autosave thread] INFO  org.bitcoinj.wallet.WalletFiles - Background saving wallet; last seen block is height 400970, date 2016-03-03T14:13:40Z, hash 00000000000000000268a135fb7456f9139272068756ea252bc444a6eef126ef
2016-04-12 15:05:56 [Wallet autosave thread] INFO  org.bitcoinj.wallet.WalletFiles - Save completed in 8.898 ms
2016-04-12 15:05:56 [NioClientManager] INFO  org.bitcoinj.wallet.WalletFiles - Saving wallet; last seen block is height 400972, date 2016-03-03T14:21:37Z, hash 000000000000000004138f6b627af6781b9d09bc17430eadb87c841a768a59fd
2016-04-12 15:05:56 [NioClientManager] INFO  org.bitcoinj.wallet.WalletFiles - Save completed in 10.40 ms
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to dnsseed.bitcoin.dashjr.org/5.249.159.54:8333
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to dnsseed.bitcoin.dashjr.org/5.249.159.54:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoin.sipa.be/82.134.66.146:8333
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoin.sipa.be/82.134.66.146:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to dnsseed.bitcoin.dashjr.org/88.212.209.106:8333
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to dnsseed.bitcoin.dashjr.org/88.212.209.106:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - [5.249.159.54]:8333: Got version=70012, subVer='/Classic:0.12.0/', services=0x5, time=2016-04-12 15:05:57, blocks=406938
2016-04-12 15:05:57 [Wallet autosave thread] INFO  org.bitcoinj.wallet.WalletFiles - Background saving wallet; last seen block is height 400994, date 2016-03-03T18:16:37Z, hash 000000000000000001d10907f05c2b7f5e12cb7351f9dd6817d1df0ba0987937
2016-04-12 15:05:57 [Wallet autosave thread] INFO  org.bitcoinj.wallet.WalletFiles - Save completed in 7.379 ms
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - [88.212.209.106]:8333: Got version=70002, subVer='/Satoshi:0.11.2/', services=0x1, time=2016-04-12 15:04:53, blocks=406938
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - [82.134.66.146]:8333: Got version=70012, subVer='/Classic:0.12.0/', services=0x5, time=2016-04-12 15:06:17, blocks=406938
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitnodes.io/76.10.188.111:8333
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitnodes.io/76.10.188.111:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Received tx for -0.04405 BTC: 62c7fc88f949e7dddc9c3f6c5d0450cdabb4516555fc66d0667fb8af9389c8da [0] in block 00000000000000000194229b35ebc23ed4cc3a3ade110e42f4ce30026a864cdd
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   marked 7664f594a906c34edc23570c3e8cdf5d0b719d2f91e2dd50c3e72de5b859b7ea:0 as spent by 62c7fc88f949e7dddc9c3f6c5d0450cdabb4516555fc66d0667fb8af9389c8da
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   7664f594a906c34edc23570c3e8cdf5d0b719d2f91e2dd50c3e72de5b859b7ea prevtx <-unspent ->spent
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   tx 62c7fc88f949e7dddc9c3f6c5d0450cdabb4516555fc66d0667fb8af9389c8da ->unspent
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Balance is now: 5.54414761 BTC
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Received tx for -0.09287 BTC: 99c08ff7dde24fed043ab6c915de6f36a58ca69ffe8dbdf59cd9c6999b7c200c [1] in block 00000000000000000194229b35ebc23ed4cc3a3ade110e42f4ce30026a864cdd
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   marked 7b5c2287d31749e4a8c57260739c30fd998d0a98c7a840d88d67a9aafd7a8ce4:0 as spent by 99c08ff7dde24fed043ab6c915de6f36a58ca69ffe8dbdf59cd9c6999b7c200c
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   7b5c2287d31749e4a8c57260739c30fd998d0a98c7a840d88d67a9aafd7a8ce4 prevtx <-unspent ->spent
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   tx 99c08ff7dde24fed043ab6c915de6f36a58ca69ffe8dbdf59cd9c6999b7c200c ->unspent
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Balance is now: 5.45127761 BTC
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Received tx for -0.06005 BTC: b2f5ce349c301fffcd97cde9aa390928ab1a5f79a07262e49c5a14e7d44d59fa [2] in block 00000000000000000194229b35ebc23ed4cc3a3ade110e42f4ce30026a864cdd
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   marked 49c275cb901332247dbb5fc90a1198a2d99e9560e9511b48cedb6f3bceb6a299:0 as spent by b2f5ce349c301fffcd97cde9aa390928ab1a5f79a07262e49c5a14e7d44d59fa
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   49c275cb901332247dbb5fc90a1198a2d99e9560e9511b48cedb6f3bceb6a299 prevtx <-unspent ->spent
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   tx b2f5ce349c301fffcd97cde9aa390928ab1a5f79a07262e49c5a14e7d44d59fa ->unspent
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Balance is now: 5.39122761 BTC
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet - Received tx for -0.04405 BTC: 62c7fc88f949e7dddc9c3f6c5d0450cdabb4516555fc66d0667fb8af9389c8da [0] in block 00000000000000000194229b35ebc23ed4cc3a3ade110e42f4ce30026a864cdd
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.wallet.Wallet -   tx 62c7fc88f949e7dddc9c3f6c5d0450cdabb4516555fc66d0667fb8af9389c8da ->unspent
2016-04-12 15:05:57 [NioClientManager] WARN  org.bitcoinj.core.PeerSocketHandler - [78.106.74.9]:8333 - 
java.lang.IllegalStateException: null
	at com.google.common.base.Preconditions.checkState(Preconditions.java:158)
	at org.bitcoinj.wallet.Wallet.addWalletTransaction(Wallet.java:2873)
	at org.bitcoinj.wallet.Wallet.processTxFromBestChain(Wallet.java:2195)
	at org.bitcoinj.wallet.Wallet.receive(Wallet.java:1947)
	at org.bitcoinj.wallet.Wallet.receiveFromBlock(Wallet.java:1887)
	at org.bitcoinj.core.AbstractBlockChain.sendTransactionsToListener(AbstractBlockChain.java:881)
	at org.bitcoinj.core.AbstractBlockChain.informListenerForNewTransactions(AbstractBlockChain.java:702)
	at org.bitcoinj.core.AbstractBlockChain.informListenersForNewBlock(AbstractBlockChain.java:627)
	at org.bitcoinj.core.AbstractBlockChain.connectBlock(AbstractBlockChain.java:567)
	at org.bitcoinj.core.AbstractBlockChain.add(AbstractBlockChain.java:489)
	at org.bitcoinj.core.AbstractBlockChain.add(AbstractBlockChain.java:381)
	at org.bitcoinj.core.BlockChain.add(BlockChain.java:166)
	at org.bitcoinj.core.Peer.endFilteredBlock(Peer.java:1081)
	at org.bitcoinj.core.Peer.processMessage(Peer.java:460)
	at org.bitcoinj.core.PeerSocketHandler.receiveBytes(PeerSocketHandler.java:182)
	at org.bitcoinj.net.ConnectionHandler.handleKey(ConnectionHandler.java:223)
	at org.bitcoinj.net.NioClientManager.handleKey(NioClientManager.java:86)
	at org.bitcoinj.net.NioClientManager.run(NioClientManager.java:122)
	at com.google.common.util.concurrent.AbstractExecutionThreadService$1$2.run(AbstractExecutionThreadService.java:60)
	at com.google.common.util.concurrent.Callables$3.run(Callables.java:95)
	at org.bitcoinj.utils.ContextPropagatingThreadFactory$1.run(ContextPropagatingThreadFactory.java:49)
	at java.lang.Thread.run(Thread.java:745)
2016-04-12 15:05:57 [NioClientManager] INFO  o.b.c.l.DownloadProgressTracker - Chain download switched to [82.134.66.146]:8333
2016-04-12 15:05:57 [NioClientManager] WARN  org.bitcoinj.net.ConnectionHandler - Error handling SelectionKey: java.lang.IllegalStateException
2016-04-12 15:05:57 [NioClientManager] INFO  org.bitcoinj.core.Peer - [76.10.188.111]:8333: Got version=70012, subVer='/Satoshi:0.12.0/', services=0x5, time=2016-04-12 15:05:57, blocks=406938
2016-04-12 15:05:58 [Wallet autosave thread] INFO  org.bitcoinj.wallet.WalletFiles - Background saving wallet; last seen block is height 400997, date 2016-03-03T18:58:01Z, hash 000000000000000002ec04eca33e2400c452b67ac0eaf9b694bbf7f10a26638d
2016-04-12 15:05:58 [Wallet autosave thread] INFO  org.bitcoinj.wallet.WalletFiles - Save completed in 7.709 ms
2016-04-12 15:05:58 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoin.sipa.be/129.21.178.73:8333
2016-04-12 15:05:58 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoin.sipa.be/129.21.178.73:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:58 [NioClientManager] INFO  org.bitcoinj.core.Peer - [129.21.178.73]:8333: Got version=70012, subVer='/Satoshi:0.12.0/', services=0x5, time=2016-04-12 15:05:58, blocks=406938
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoinstats.com/188.138.102.42:8333
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoinstats.com/188.138.102.42:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to dnsseed.bitcoin.dashjr.org/89.46.74.78:8333
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to dnsseed.bitcoin.dashjr.org/89.46.74.78:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.core.Peer - [188.138.102.42]:8333: Got version=70002, subVer='/Satoshi:0.9.4/', services=0x1, time=2016-04-12 15:06:28, blocks=406938
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.core.Peer - [89.46.74.78]:8333: Got version=70012, subVer='/Classic:0.12.0/', services=0x5, time=2016-04-12 15:06:00, blocks=406938
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoinstats.com/119.81.11.69:8333
2016-04-12 15:05:59 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoinstats.com/119.81.11.69:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:00 [NioClientManager] INFO  org.bitcoinj.core.Peer - [119.81.11.69]:8333: Got version=70012, subVer='/Classic:0.12.0/', services=0x5, time=2016-04-12 15:05:39, blocks=406938
2016-04-12 15:06:00 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoin.sipa.be/75.143.109.42:8333
2016-04-12 15:06:00 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoin.sipa.be/75.143.109.42:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:01 [NioClientManager] INFO  org.bitcoinj.core.Peer - [75.143.109.42]:8333: Got version=70012, subVer='/Classic:0.12.0/', services=0x5, time=2016-04-12 15:06:01, blocks=406938
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoinstats.com/136.243.124.142:8333
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoinstats.com/136.243.124.142:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoinstats.com/136.243.13.76:8333
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoinstats.com/136.243.13.76:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.core.Peer - [136.243.124.142]:8333: Got version=70002, subVer='/Classic:0.11.2/', services=0x1, time=2016-04-12 15:02:33, blocks=406938
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.core.Peer - [136.243.13.76]:8333: Got version=70002, subVer='/Satoshi:0.11.2/', services=0x1, time=2016-04-12 15:06:02, blocks=406938
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoin.sipa.be/67.205.59.222:8333
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoin.sipa.be/67.205.59.222:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoin.sipa.be/113.148.158.231:8333
2016-04-12 15:06:02 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoin.sipa.be/113.148.158.231:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:03 [NioClientManager] INFO  org.bitcoinj.core.Peer - [67.205.59.222]:8333: Got version=70011, subVer='/Bitcoin XT:0.11.0E(Linux; x86_64)/', services=0x7, time=2016-04-12 15:05:56, blocks=406938
2016-04-12 15:06:03 [NioClientManager] WARN  org.bitcoinj.core.Peer - [67.205.59.222]:8333: Received unhandled message: Bloom Filter of size 0 with 0 hash functions.
2016-04-12 15:06:03 [NioClientManager] INFO  org.bitcoinj.core.Peer - [113.148.158.231]:8333: Got version=70002, subVer='/Satoshi:0.11.0/', services=0x1, time=2016-04-12 15:04:38, blocks=406938
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to dnsseed.bitcoin.dashjr.org/85.214.124.168:8333
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to dnsseed.bitcoin.dashjr.org/85.214.124.168:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.net.NioClientManager - Connected to seed.bitcoinstats.com/89.38.150.89:8333
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.core.Peer - Announcing to seed.bitcoinstats.com/89.38.150.89:8333 as: /bitcoinj:0.15-SNAPSHOT/
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.core.Peer - [85.214.124.168]:8333: Got version=70002, subVer='/Satoshi:0.11.2/', services=0x1, time=2016-04-12 15:05:39, blocks=406938
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.core.Peer - [89.38.150.89]:8333: Got version=70012, subVer='/Classic:0.12.0/', services=0x5, time=2016-04-12 15:06:05, blocks=406938
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.wallet.WalletFiles - Saving wallet; last seen block is height 400999, date 2016-03-03T19:02:46Z, hash 0000000000000000063829b985b700dbd4963776a7d3fd64452e89cca1111b46
2016-04-12 15:06:04 [NioClientManager] INFO  org.bitcoinj.wallet.WalletFiles - Save completed in 9.491 ms

**Comment by PavelNiedoba** [comment]
Full log since wallet restore from seed

[fullLog.txt](https://github.com/bitcoinj/bitcoinj/files/215060/fullLog.txt)
```java
fragments of wallet creation:
 log.info("Create wallet from seed");
                params = Context.get().getNetworkParameters();
                wallet = Wallet.fromSeed(params, BitcoinTools.getDeterministicSeed(
                        Context.get().getSeed(),
                        Context.get().getSeedCreationDate()
                ));
                wallet.saveToFile(walletFile);
wallet.autosaveToFile(walletFile, 1, TimeUnit.SECONDS, null);

 blockStore = new SPVBlockStore(params, blockStoreFile);

            chain = new BlockChain(params, wallet, blockStore);

            peerGroup = new PeerGroup(params, chain);
            chain.addWallet(wallet);
            peerGroup.addWallet(wallet);
            peerGroup.addPeerDiscovery(new DnsDiscovery(params));

            peerGroup.start();
            peerGroup.startBlockChainDownload(bListener);

```

**Comment by PavelNiedoba** [comment]
Fast debug shows situation like this:
Blockchain is downloading, it finds transaction and tries to put it into map of unspent transactions. 

```java
   private void addWalletTransaction(Pool pool, Transaction tx) {
        checkState(lock.isHeldByCurrentThread());
        transactions.put(tx.getHash(), tx);
        switch (pool) {
        case UNSPENT:
            checkState(unspent.put(tx.getHash(), tx) == null);
            break;
        case SPENT:
...
```

But it's already here.  I have no idea why this could happen.

**Comment by schildbach** [comment]
If you replay the blockchain, do you use `Wallet.reset()` to clear all transactions from the wallet?

I suggest subclassing `org.bitcoinj.kits.WalletAppKit` or at least have a look at that source for inspiration.

**Comment by PavelNiedoba** [comment]
To replay blockchain I delete blockstore and wallet files. I need to know that I can fully recover wallet from seed only. 

More over, I removed  checkState( ) part,  which just throws exception in case transaction was already downloaded and closes peer which leads to situation I'm out of peers after few blocks and blockchain download stops. 

whole method looks like:
``` java
  /**
     * Adds the given transaction to the given pools and registers a confidence change listener on it.
     */
    private void addWalletTransaction(Pool pool, Transaction tx) {
        checkState(lock.isHeldByCurrentThread());
        transactions.put(tx.getHash(), tx);
        switch (pool) {
        case UNSPENT:
            if (unspent.put(tx.getHash(), tx) == null) {
                log.error("transaction {} added to {} but its already there", tx.getHash(), pool);
            }
            break;
        case SPENT:
            checkState(spent.put(tx.getHash(), tx) == null);
            break;
        case PENDING:
            checkState(pending.put(tx.getHash(), tx) == null);
            break;
        case DEAD:
            checkState(dead.put(tx.getHash(), tx) == null);
            break;
        default:
            throw new RuntimeException("Unknown wallet transaction type " + pool);
        }

        if (pool == Pool.UNSPENT || pool == Pool.PENDING)

        {
            for (TransactionOutput output : tx.getOutputs()) {
                if (output.isAvailableForSpending() && output.isMineOrWatched(this))
                    myUnspents.add(output);
            }
        }
        // This is safe even if the listener has been added before, as TransactionConfidence ignores duplicate
        // registration requests. That makes the code in the wallet simpler.
        tx.getConfidence().

                addEventListener(Threading.SAME_THREAD, txConfidenceListener);
    }
```

After that blockchain downloaded on first try and also much faster, because peers are not switching at all. 

I also logged duplicated transactions and it seems like every transaction is added twice. 

Please consider this as potential bug. I guess it will not affect most users, since they don't have 1000+ transactions and they restart wallet regularly. 

I have no idea if this has security impacts.

**Comment by schildbach** [comment]
If you believe it's dependent on your wallet I would need some way to reproduce the issue. One way would be you post the "xpub" and the creation date. This would disclose all addresses and transactions of your wallet, but not the private keys. The wallet could be replayed as a watch-only wallet and hopefully the issue reproduced.

You could also try this yourself in wallet-tool:

```
./wallet-tool create --wallet=your.wallet --watchkey=xpub… --date=yyyy/mm/dd```
./wallet-tool sync --debug --wallet=your.wallet
```

**Comment by PavelNiedoba** [comment]
Watching key 023653c4584e4f2687f143cd4cd3943acb92837b33f8f9948ec884d5cfdbe25a8d
and 
xpub688i2h6a8ZCdQQx14hESse2WSe7opBhDm3osoWMQJa7jtTL2D3LjTgXp9K13hQ5PFfF36Bp6tgSYBLKDkwvjKtztouaQhprFU3rybJfttTQ

**Comment by schildbach** [comment]
I was able to replay your wallet without problems with the bitcoinj on current master; it determined a balance of 2.29844413 BTC.

I suspect you didn't publish all your code and the changes you did to bitcoinj. For example, removing the checkState() is bad idea – transactions may not enter the wallet twice.

**Comment by schildbach** [comment]
Pavel, did you manage to solve your issue?

**Comment by PavelNiedoba** [comment]
Andreas

  Yes, I managed to develop stable wallet service. It include the hack I pasted here and I also wrote watchdog thread, which regularly checks peers in peergroup for pending pings and slow pings. 
  I have 5 weeks uptime now. Before I had 5 days maximum. Suppressing exception when adding transaction doesn't seem to make any problems. 
  I modified bitcoinj library a bit. Only opened access to peergroup so I can have the watchdog  looking into it.
